piptools/repositories/pypi.py

```diff
@@ -91,7 +91,7 @@ def find_best_match(self, ireq, prereleases=None):
         Returns a Version object that indicates the best match for the given
         InstallRequirement according to the external repository.
         """
-        if ireq.editable:
+        if ireq.editable or (ireq.link and not ireq.link.is_artifact):
```

Comment:
Then you can also make another helper method `is_editable_vcs_link(ireq)`, which also does the `.editable` check. I think having those functions will make the rest of the changes below more readable.